sdks/java/extensions/smb/src/main/java/org/apache/beam/sdk/extensions/smb/SortedBucketSource.java

```diff
@@ -0,0 +1,304 @@
+import org.apache.beam.sdk.io.FileSystems;
+import org.apache.beam.sdk.io.fs.ResourceId;
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.Reshuffle;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PValue;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.beam.vendor.guava.v20_0.com.google.common.base.Preconditions;
+
+/**
+ * Reads in an arbitrary number of heterogeneous typed Sources stored with the same bucketing
+ * scheme, and co-groups on a common key.
+ *
+ * @param <KeyT>
+ * @param <ResultT>
+ */
+public class SortedBucketSource<KeyT, ResultT>
+    extends PTransform<PBegin, PCollection<KV<KeyT, ResultT>>> {
+  private final Coder<KeyT> keyCoder;
+  private final SMBCoGbkResult.ToResult<ResultT> toResult;
+  private final List<KeyedBucketSource<KeyT, ?>> sources;
+  private final Comparator<KeyT> keyComparator;
+
+  public SortedBucketSource(
+      Coder<KeyT> keyCoder,
+      SMBCoGbkResult.ToResult<ResultT> toResult,
+      List<KeyedBucketSource<KeyT, ?>> sources,
+      Comparator<KeyT> keyComparator) {
+    this.keyCoder = keyCoder;
+    this.sources = sources;
+    this.toResult = toResult;
+    this.keyComparator = keyComparator;
+  }
+
+  @Override
+  public final PCollection<KV<KeyT, ResultT>> expand(PBegin begin) {
+
+    // @TODO: Support asymmetric, but still compatible, bucket sizes in reader.
+
+    BucketMetadata<KeyT, ?> first = null;
+    for (KeyedBucketSource<KeyT, ?> source : sources) {
+      BucketMetadata<KeyT, ?> current = source.readMetadata();
+      if (first == null) {
+        first = current;
+      } else {
+        Preconditions.checkState(first.compatibleWith(current));
+      }
+    }
+
+    final int numBuckets = first.getNumBuckets();
+
+    final PCollection<KV<Integer, List<BucketSourceIterator<KeyT>>>> openedReaders =
+        (PCollection<KV<Integer, List<BucketSourceIterator<KeyT>>>>)
+            new KeyedBucketSources<>(begin.getPipeline(), numBuckets, sources)
+                .expand()
+                .get(new TupleTag<>("readers"));
+
+    return openedReaders
+        .apply("Force each key to a separate core", Reshuffle.viaRandomKey())
+        .apply(
+            "Perform co-group operation per key",
+            ParDo.of(new MergeBuckets<>(toResult, keyComparator)))
+        .setCoder(KvCoder.of(keyCoder, toResult.resultCoder()));
+  }
+
+  /** @param <KeyT> */
+  static class MergeBuckets<KeyT, ResultT>
+      extends DoFn<KV<Integer, List<BucketSourceIterator<KeyT>>>, KV<KeyT, ResultT>> {
+    private final SMBCoGbkResult.ToResult<ResultT> toResult;
+    private final Comparator<KeyT> keyComparator;
+
+    MergeBuckets(ToResult<ResultT> toResult, Comparator<KeyT> keyComparator) {
```

Comment:
We can get rid of `keyComparator` here, compare bytes, and decode to `KeyT` (unified key type) before outputing.